File: 12. 整数转罗马数字.js
/**
 * @param {number} num
 * @return {string}
 */
var intToRoman = function(num) {
    let out = '';
    const table = [
        {str: 'M', value: 1000},
        {str: 'CM', value: 900},
        {str: 'D', value: 500},
        {str: 'CD', value: 400},
        {str: 'C', value: 100},
        {str: 'XC', value: 90},
        {str: 'L', value: 50},
        {str: 'XL', value: 40},
        {str: 'X', value: 10},
        {str: 'IX', value: 9},
        {str: 'V', value: 5},
        {str: 'IV', value: 4},
        {str: 'I', value: 1}
    ];

    for (let i = 0; i < table.length; i++) {
        let n = parseInt(num / table[i].value);
        while (n--) out += table[i].str;
        num %= table[i].value;
    }

    return out;
};

File: 12. 整数转罗马数字.cpp
class Solution {
public:
    string maxNumToString(int &originNum, int reduceNum, string target) {
        string s = "";
        while (originNum >= reduceNum) {
            originNum -= reduceNum;
            s += target;
        }
        return s;
    }
    string intToRoman(int num) {
        vector<pair<int, string>> v = {
            {1000, "M"},
            {900, "CM"},
            {500, "D"},
            {400, "CD"},
            {100, "C"},
            {90, "XC"},
            {50, "L"},
            {40, "XL"},
            {10, "X"},
            {9, "IX"},
            {5, "V"},
            {4, "IV"},
            {1, "I"}
        };
        string s = "";

        for (auto p : v) {
            s += maxNumToString(num, p.first, p.second);
        }

        return s;
    }
};